tests/integration/test_with_portainer_service.py

```diff
@@ -21,9 +65,230 @@ async def test_portainer_connection(
     event_loop: asyncio.AbstractEventLoop,
     portainer_container: tuple[URL, str],
     aiohttp_client_session: ClientSession,
-):
+) -> str:
     portainer_url, portainer_password = portainer_container
 
-    await portainer.authenticate(
+    return await portainer.authenticate(
         portainer_url, aiohttp_client_session, "admin", portainer_password
     )
+
+
+@pytest.fixture
+async def portainer_endpoint_id(
+    loop: asyncio.AbstractEventLoop,
+    portainer_container: tuple[URL, str],
+    aiohttp_client_session: ClientSession,
+    portainer_baerer_code: str,
+) -> int:
+    portainer_url, _ = portainer_container
+
+    endpoint = await portainer.get_first_endpoint_id(
+        portainer_url, aiohttp_client_session, portainer_baerer_code
+    )
+    assert type(endpoint) == int
+    return endpoint
+
+
+async def test_portainer_test_create_stack(
+    loop: asyncio.AbstractEventLoop,
+    portainer_container: tuple[URL, str],
+    aiohttp_client_session: ClientSession,
+    portainer_baerer_code: str,
+    portainer_endpoint_id: int,
+    valid_docker_stack,
+    docker_swarm: None,
+):
+    portainer_url, _ = portainer_container
+    try_to_format_url = URL(portainer_url)
+    print(try_to_format_url)
+    swarm_id = await portainer.get_swarm_id(
+        portainer_url,
+        aiohttp_client_session,
+        portainer_baerer_code,
+        portainer_endpoint_id,
+    )
+    current_stack_name = "pytestintegration"
+    # Assuring a clean state by deleting any remnants
+    os.system("docker stack rm " + current_stack_name)
+
+    new_stack = await portainer.post_new_stack(
+        base_url=portainer_url,
+        app_session=aiohttp_client_session,
+        bearer_code=portainer_baerer_code,
+        swarm_id=swarm_id,
+        endpoint_id=portainer_endpoint_id,
+        stack_name=current_stack_name,
+        stack_cfg=valid_docker_stack,
+    )
+    time.sleep(2)
```

Comment:
I see lots of sleeps in the code. If you are waiting for something to be ready, I use a retry pattern with tenacity. This makes it less likely for you test to fail as flaky.

To do so you make a call to something that should respond/ be there, if there is no error you proceed otherwise you retry every X seconds for a maximum of Y seconds.

This pattern is more robust than simply sleeping, in simcore were are trying to remove these fixed sleeps, that can cause issues.

Reply:
adressed, less sleeps now